core/shared/src/main/scala/zio/stm/TReentrantLock.scala

```diff
@@ -98,32 +108,28 @@ final class TReentrantLock private (data: TRef[Either[ReadLock, WriteLock]]) {
    * Releases a read lock held by this fiber. Succeeds with the outstanding
    * number of read locks held by this fiber.
    */
-  lazy val releaseRead: USTM[Int] =
-    STM.fiberId.flatMap(fiberId => adjustRead(fiberId, -1))
+  lazy val releaseRead: USTM[Unit] = adjustRead(-1)
 
   /**
    * Releases a write lock held by this fiber. Succeeds with the outstanding
    * number of write locks held by this fiber.
    */
-  lazy val releaseWrite: USTM[Int] =
-    STM.fiberId.flatMap(fiberId =>
-      data.modify {
-        case Right(WriteLock(1, m, `fiberId`)) =>
-          0 -> Left(ReadLock(fiberId, m))
-
-        case Right(WriteLock(n, m, `fiberId`)) if n > 1 =>
-          val newCount = n - 1
-
-          newCount -> Right(WriteLock(newCount, m, fiberId))
-
-        case s => die(s"Defect: Fiber ${fiberId} releasing write lock it does not hold: ${s}")
-      }
-    )
+  lazy val releaseWrite: USTM[Unit] = new ZSTM((journal, fiberId, _, _) => {
+    val res = data.unsafeGet(journal) match {
+      case Right(WriteLock(1, m, `fiberId`)) => Left(ReadLock(fiberId, m))
```

Comment:
There's an optimization you can do here, if you want: right now we represent the state as "either". But we can fuse that sum type into a new type `LockState`, which will be either `WriteLock` or `ReadLock`. This will eliminate 1 layer of wrapping and a bunch of needless allocations.